File: mobile_app/test/admin_users_role_filter_navigation_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:mobile_app/core/routes/app_routes.dart';
import 'package:mobile_app/features/dashboard/data/admin_users_repository.dart';
import 'package:mobile_app/features/dashboard/presentation/admin/admin_users_screen.dart';
import 'package:mobile_app/l10n/app_localizations.dart';

Future<AppLocalizations> _loadLocalizations(Locale locale) async {
  return lookupAppLocalizations(locale);
}

bool _matchesRole(AdminUserRecord user, String? roleFilter) {
  return roleFilter == null ||
      roleFilter.isEmpty ||
      user.role.toLowerCase() == roleFilter.toLowerCase();
}

List<AdminUserRecord> _filterUsersByRole(
  List<AdminUserRecord> users,
  String? roleFilter,
) {
  return users
      .where((user) => _matchesRole(user, roleFilter))
      .toList(growable: false);
}

void main() {
  group('Admin Users role filter navigation', () {
    late AppLocalizations en;
    late AppLocalizations ar;

    setUpAll(() async {
      en = await _loadLocalizations(const Locale('en'));
      ar = await _loadLocalizations(const Locale('ar'));
    });

    test('Specialists KPI route includes role=specialist', () {
      expect(
        AppRoutes.adminUsersWithRole(role: 'specialist'),
        '${AppRoutes.adminUsers}?role=specialist',
      );
    });

    test('plain Users route has no role query param', () {
      expect(AppRoutes.adminUsersWithRole(), AppRoutes.adminUsers);
      expect(AppRoutes.adminUsersWithRole(role: null), AppRoutes.adminUsers);
    });

    test('parseAdminUsersRoleParam accepts specialist and rejects invalid values', () {
      expect(AppRoutes.parseAdminUsersRoleParam('specialist'), 'specialist');
      expect(AppRoutes.parseAdminUsersRoleParam(' SPECIALIST '), 'specialist');
      expect(AppRoutes.parseAdminUsersRoleParam(null), isNull);
      expect(AppRoutes.parseAdminUsersRoleParam(''), isNull);
      expect(AppRoutes.parseAdminUsersRoleParam('all'), isNull);
      expect(AppRoutes.parseAdminUsersRoleParam('unknown'), isNull);
      expect(AppRoutes.parseAdminUsersRoleParam('الأخصائي'), isNull);
    });

    test('AdminUsersScreen stores parsed initial role filter once', () {
      const screen = AdminUsersScreen(initialRoleFilter: 'specialist');
      expect(screen.initialRoleFilter, 'specialist');

      const defaultScreen = AdminUsersScreen();
      expect(defaultScreen.initialRoleFilter, isNull);
    });

    test('existing role filtering returns only specialists for specialist filter', () {
      final users = [
        AdminUserRecord(
          id: '1',
          name: 'Admin User',
          email: 'admin@test.com',
          role: 'admin',
          isActive: true,
        ),
        AdminUserRecord(
          id: '2',
          name: 'Spec User',
          email: 'spec@test.com',
          role: 'specialist',
          isActive: true,
        ),
        AdminUserRecord(
          id: '3',
          name: 'Parent User',
          email: 'parent@test.com',
          role: 'parent',
          isActive: true,
        ),
      ];

      final filtered = _filterUsersByRole(users, 'specialist');

      expect(filtered, hasLength(1));
      expect(filtered.first.role, 'specialist');
    });

    test('missing or invalid role defaults to All via null filter', () {
      expect(AppRoutes.parseAdminUsersRoleParam(null), isNull);
      expect(AppRoutes.parseAdminUsersRoleParam('invalid'), isNull);

      final users = [
        AdminUserRecord(
          id: '1',
          name: 'Admin User',
          email: 'admin@test.com',
          role: 'admin',
          isActive: true,
        ),
        AdminUserRecord(
          id: '2',
          name: 'Spec User',
          email: 'spec@test.com',
          role: 'specialist',
          isActive: true,
        ),
      ];

      expect(_filterUsersByRole(users, null), users);
    });

    test('role filter value is locale-independent while labels differ', () {
      expect(en.roleSpecialist, isNot(equals(ar.roleSpecialist)));
      expect(AppRoutes.parseAdminUsersRoleParam('specialist'), 'specialist');
    });

    test('manual filter switching uses the same raw role values', () {
      String? roleFilter = AppRoutes.parseAdminUsersRoleParam('specialist');
      expect(roleFilter, 'specialist');

      roleFilter = null;
      expect(roleFilter, isNull);

      roleFilter = AppRoutes.parseAdminUsersRoleParam('parent');
      expect(roleFilter, 'parent');
    });
  });
}

File: frontend_web/src/features/admin-dashboard/utils/adminUsersRoleFilterNavigation.test.js
import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { createTranslator } from "../../../i18n/index.js";
import {
  ADMIN_WEB_ROUTES,
  buildAdminUsersPath,
  parseAdminUsersRoleParam,
} from "../../../routes/adminDashboardRoutes.js";
import { filterAdminUsers } from "./adminUsersMappers.js";
import { getAdminUsersRoleFilterOptions } from "./adminUsersLocalization.js";

describe("adminUsersRoleFilterNavigation", () => {
  it("builds the Users route with role=specialist for the Specialists KPI deep link", () => {
    assert.equal(
      buildAdminUsersPath("specialist"),
      `${ADMIN_WEB_ROUTES.users}?role=specialist`,
    );
  });

  it("builds the plain Users route when no role is provided", () => {
    assert.equal(buildAdminUsersPath(null), ADMIN_WEB_ROUTES.users);
    assert.equal(buildAdminUsersPath(undefined), ADMIN_WEB_ROUTES.users);
    assert.equal(buildAdminUsersPath(""), ADMIN_WEB_ROUTES.users);
  });

  it("parses role=specialist into the specialist raw filter value", () => {
    assert.equal(parseAdminUsersRoleParam("specialist"), "specialist");
    assert.equal(parseAdminUsersRoleParam("SPECIALIST"), "specialist");
    assert.equal(parseAdminUsersRoleParam(" specialist "), "specialist");
  });

  it("defaults to All when the role query param is missing or invalid", () => {
    assert.equal(parseAdminUsersRoleParam(null), null);
    assert.equal(parseAdminUsersRoleParam(undefined), null);
    assert.equal(parseAdminUsersRoleParam(""), null);
    assert.equal(parseAdminUsersRoleParam("all"), null);
    assert.equal(parseAdminUsersRoleParam("unknown"), null);
    assert.equal(parseAdminUsersRoleParam("الأخصائي"), null);
  });

  it("accepts only the existing supported role filter values", () => {
    assert.equal(parseAdminUsersRoleParam("admin"), "admin");
    assert.equal(parseAdminUsersRoleParam("parent"), "parent");
  });

  it("filters users to specialists using the existing role filter logic", () => {
    const users = [
      { id: "1", role: "admin", fullName: "Admin User", email: "admin@test.com" },
      { id: "2", role: "specialist", fullName: "Spec User", email: "spec@test.com" },
      { id: "3", role: "parent", fullName: "Parent User", email: "parent@test.com" },
    ];

    const filtered = filterAdminUsers(users, { roleFilter: "specialist" });

    assert.equal(filtered.length, 1);
    assert.equal(filtered[0].role, "specialist");
  });

  it("keeps role filter values locale-independent while labels differ by locale", () => {
    const translateEn = createTranslator("en");
    const translateAr = createTranslator("ar");
    const enFilters = getAdminUsersRoleFilterOptions(translateEn);
    const arFilters = getAdminUsersRoleFilterOptions(translateAr);
    const enSpecialist = enFilters.find((filter) => filter.id === "specialist");
    const arSpecialist = arFilters.find((filter) => filter.id === "specialist");

    assert.equal(enSpecialist.value, "specialist");
    assert.equal(arSpecialist.value, "specialist");
    assert.notEqual(enSpecialist.label, arSpecialist.label);
    assert.equal(parseAdminUsersRoleParam(enSpecialist.value), "specialist");
    assert.equal(parseAdminUsersRoleParam(arSpecialist.value), "specialist");
  });
});
